File: app/src/main/java/com/hippo/ehviewer/client/data/TagSet.java
/*
 * Copyright 2017 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hippo.ehviewer.client.data;

/*
 * Created by Hippo on 5/15/2017.
 */

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

/**
 * A set which stores multiple tags.
 */
public final class TagSet implements Iterable<Map.Entry<String, Set<String>>>, Parcelable {

  private Map<String, Set<String>> groups = new HashMap<>();

  /**
   * Put a tag to this set.
   */
  public void add(@NonNull String namespace, @NonNull String tag) {
    Set<String> set = groups.get(namespace);
    if (set == null) {
      set = new HashSet<>();
      groups.put(namespace, set);
    }
    // Don't add duplicate tag
    if (!set.contains(tag)) {
      set.add(tag);
    }
  }

  /**
   * Remove a tag from this set.
   */
  public void remove(@NonNull String namespace, @NonNull String tag) {
    Set<String> set = groups.get(namespace);
    if (set != null) {
      set.remove(tag);
      // Remove the namespace if it's empty
      if (set.isEmpty()) {
        groups.remove(namespace);
      }
    }
  }

  public void clear() {
    groups.clear();
  }

  /**
   * Make this set the same as another set.
   */
  public void set(@NonNull TagSet tagSet) {
    groups.clear();
    for (Map.Entry<String, Set<String>> entry : tagSet) {
      groups.put(entry.getKey(), new HashSet<>(entry.getValue()));
    }
  }

  /**
   * Returns the size of tags.
   */
  public int size() {
    int sum = 0;
    for (Set<String> set : groups.values()) {
      sum += set.size();
    }
    return sum;
  }

  /**
   * Returns {@code true} if no tags.
   */
  public boolean isEmpty() {
    return groups.isEmpty();
  }

  @Override
  public Iterator<Map.Entry<String, Set<String>>> iterator() {
    return groups.entrySet().iterator();
  }

  @Override
  public boolean equals(Object obj) {
    return obj instanceof TagSet && groups.equals(((TagSet) obj).groups);
  }

  @Override
  public int hashCode() {
    return groups.hashCode();
  }

  @Override
  public String toString() {
    return "TagSet: " + groups.toString();
  }

  @Override
  public int describeContents() {
    return 0;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    dest.writeInt(groups.size());
    for (Map.Entry<String, Set<String>> entry : groups.entrySet()) {
      dest.writeString(entry.getKey());

      Set<String> set = entry.getValue();
      dest.writeInt(set.size());
      for (String tag: set) {
        dest.writeString(tag);
      }
    }
  }

  public TagSet() {}

  protected TagSet(Parcel in) {
    int groupSize = in.readInt();
    groups = new HashMap<>(groupSize);
    for (int i = 0; i < groupSize; ++i) {
      String key = in.readString();

      int tagSize = in.readInt();
      Set<String> set = new HashSet<>(tagSize);
      for (int j = 0; j < tagSize; ++j) {
        set.add(in.readString());
      }
      groups.put(key, set);
    }
  }

  public static final Parcelable.Creator<TagSet> CREATOR = new Parcelable.Creator<TagSet>() {
    @Override
    public TagSet createFromParcel(Parcel source) {
      return new TagSet(source);
    }

    @Override
    public TagSet[] newArray(int size) {
      return new TagSet[size];
    }
  };
}

File: app/src/main/java/com/hippo/ehviewer/client/data/TagSet.kt
/*
 * Copyright 2017 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hippo.ehviewer.client.data

import android.os.Parcel
import android.os.Parcelable

/*
 * Created by Hippo on 2017/7/25.
 */

class TagSet() : Iterable<Map.Entry<String, List<String>>>, Parcelable {

  private val map = LinkedHashMap<String, MutableList<String>>()

  fun add(namespace: String, tag: String) {
    var list = map[namespace]
    if (list == null) {
      list = arrayListOf()
      map[namespace] = list
    }
    list.add(tag)
  }

  fun remove(namespace: String, tag: String) {
    val list = map[namespace]
    if (list != null) {
      list.remove(tag)
      if (list.isEmpty()) {
        map.remove(namespace)
      }
    }
  }

  override fun iterator(): Iterator<Map.Entry<String, List<String>>> = map.iterator()

  fun size(): Int = map.values.sumBy { it.size }

  fun isEmpty(): Boolean = map.isEmpty()

  fun firstTag(): Pair<String, String>? {
    val iterator = map.iterator()
    if (iterator.hasNext()) {
      val (namespace, tags) = iterator.next()
      if (tags.isNotEmpty()) {
        return Pair(namespace, tags[0])
      } else {
        return null
      }
    } else {
      return null
    }
  }

  fun set(tags: TagSet) {
    map.clear()
    for ((namespace, list) in tags) {
      map[namespace] = ArrayList(list)
    }
  }

  constructor(parcel: Parcel) : this() {
    for (i in 0 until parcel.readInt()) {
      val namespace = parcel.readString()
      val list = arrayListOf<String>()
      map[namespace] = list

      for (j in 0 until parcel.readInt()) {
        list.add(parcel.readString())
      }
    }
  }

  override fun writeToParcel(parcel: Parcel, flags: Int) {
    parcel.writeInt(map.size)

    for ((namespace, list) in map) {
      parcel.writeString(namespace)

      parcel.writeInt(list.size)
      for (tag in list) {
        parcel.writeString(tag)
      }
    }
  }

  override fun describeContents(): Int = 0

  companion object CREATOR : Parcelable.Creator<TagSet> {
    override fun createFromParcel(parcel: Parcel): TagSet {
      return TagSet(parcel)
    }

    override fun newArray(size: Int): Array<TagSet?> {
      return arrayOfNulls(size)
    }
  }
}
